phabricator.js

```diff
@@ -0,0 +1,157 @@
+
+let parentRevision = null;
+let failedToFetch = false;
+
+async function injectButtons() {
+  let blocks = [];
+  let oldCount = 0;
+  do {
+    await wait(1000);
+    oldCount = blocks.length;
+    blocks = document.querySelectorAll('div[data-sigil=differential-changeset]');
+  } while(!blocks.length || oldCount != blocks.length);
+
+  blocks.forEach(block => {
+    let path = block.querySelector('h1.differential-file-icon-header').textContent;
+    let buttonDiv = block.querySelector('div.differential-changeset-buttons');
+    let lines = block.querySelectorAll('table.differential-diff tbody tr th:first-child');
+
+    let button = document.createElement('button');
+    button.type = 'button';
+    button.textContent = 'Code Coverage';
+    button.className = 'button button-grey';
+
+    let enabled = false;
+    let data = null;
+    async function toggle() {
+      enabled = !enabled;
+      if (enabled) {
+        if (data == null) {
+          while (!parentRevision && !failedToFetch) {
+            await wait(1000);
+          }
+          if (failedToFetch) {
+            // TODO signal failure to the user
+            console.error('Failed to fetch parent revision.');
+            enabled = false;
+            button.disabled = true;
+            button.textContent = 'Error retrieving Code Coverage';
+            return;
+          }
+
+          let coverageData = await fetchCoverage(parentRevision, path);
+          if (!coverageData.hasOwnProperty('data')) {
+            // TODO signal to user that there is no coverage data
+            console.log("The file " + path + " has no coverage data.");
+            button.disabled = true;
+            button.textContent = 'No Code Coverage Data'
+            enabled = false;
+            return;
+          }
+          data = coverageData.data;
+        }
+
+        for (let line of lines) {
+          let lineNumber = parseInt(line.textContent);
+          if (isNaN(lineNumber)) {
+            continue;
+          }
+          if (!data.hasOwnProperty(lineNumber)) {
+            continue;
+          }
+
+          if (data[lineNumber] > 0) {
+            line.style.backgroundColor = 'greenyellow';
+          } else {
+            line.style.backgroundColor = 'tomato';
+          }
+        }
+      } else {
+        for (let line of lines) {
+          line.style.backgroundColor = '';
+        }
+      }
+    }
+    button.onclick = toggle;
+
+    buttonDiv.append(button);
+  });
+}
+
+
+function getParentRevision() {
+  const querydiffsUrl = 'https://phabricator.services.mozilla.com/api/differential.querydiffs';
+  const diffIdPattern = RegExp('/D(\\d+)');
+
+  let urlPathname = document.location.pathname;
+  let diffIdMatch = urlPathname.match(diffIdPattern);
+  if (!diffIdMatch) {
+    console.error('diff id not found! urlPathname = ' + urlPathname);
+    return new Promise(r => { r(null); });
+  }
+  let diffId = diffIdMatch[1];
+
+  return new Promise(async (resolve) => {
```

Comment:
Make getParentRevision an async function, so you don't have to manually return a promise